concordium-consensus/src/Concordium/GlobalState/Basic/TreeState.hs

```diff
@@ -238,38 +246,46 @@ instance (bs ~ BlockState m, BS.BlockStateStorage m, Monad m, MonadIO m, MonadSt
             Nothing -> Map.toAscList beyond
             Just s -> (sn, s) : Map.toAscList beyond
 
-    addCommitTransaction bi@WithMetadata{..} slot = do
+    addCommitTransaction ((bi@WithMetadata{..}, verResCtx), ts) slot = do
       let trHash = wmdHash
       tt <- use transactionTable
       case tt ^. ttHashMap . at' trHash of
-          Nothing ->
-            case wmdData of
-              NormalTransaction tr -> do
-                let sender = accountAddressEmbed (transactionSender tr)
-                    nonce = transactionNonce tr
-                if (tt ^. ttNonFinalizedTransactions . at' sender . non emptyANFT . anftNextNonce) <= nonce then do
-                  transactionTablePurgeCounter += 1
-                  let wmdtr = WithMetadata{wmdData=tr,..}
-                  transactionTable .= (tt & (ttNonFinalizedTransactions . at' sender . non emptyANFT . anftMap . at' nonce . non Set.empty %~ Set.insert wmdtr)
-                                          & (ttHashMap . at' trHash ?~ (bi, Received slot)))
-                  return (TS.Added bi)
-                else return TS.ObsoleteNonce
-              CredentialDeployment{} -> do
-                transactionTable . ttHashMap . at' trHash ?= (bi, Received slot)
-                return (TS.Added bi)
-              ChainUpdate cu -> do
-                let uty = updateType (uiPayload cu)
-                    sn = updateSeqNumber (uiHeader cu)
-                if (tt ^. ttNonFinalizedChainUpdates . at' uty . non emptyNFCU . nfcuNextSequenceNumber) <= sn then do
-                  transactionTablePurgeCounter += 1
-                  let wmdcu = WithMetadata{wmdData=cu,..}
-                  transactionTable .= (tt
-                          & (ttNonFinalizedChainUpdates . at' uty . non emptyNFCU . nfcuMap . at' sn . non Set.empty %~ Set.insert wmdcu)
-                          & (ttHashMap . at' trHash ?~ (bi, Received slot)))
-                  return (TS.Added bi)
-                else return TS.ObsoleteNonce
-          Just (_, Finalized{}) ->
-            return TS.ObsoleteNonce
+          Nothing -> do
+            -- If the transaction was not present in the `TransactionTable` then we verify it now
+            -- and adding it based on the verification result.
+            -- Only transactions which can possible be valid at the stage of execution are being added
+            -- to the `TransactionTable`.
+            -- Verifying the transaction here as opposed to `doReceiveTransactionInternal` avoids
+            -- verifying a transaction which have been both received individually and as part of a block twice.
```

Comment:
I see. 

A little explanation. 

Originally there was some verification process in `doReceiveTransactionInternal`. But this would result in transactions being verified twice. So I moved it to the `addCommitTransaction`.  (When they were received individually and as part of a block). 

So I desecrated the abstraction boundaries a bit with the goal in mind of a 50% reduction in performance.